docs/design/2020-08-04-global-index.md

```diff
@@ -0,0 +1,191 @@
+
+```
+Key: tablePrefix{tableID}_indexPrefixSep{indexID}_indexedColumnsValue
+Value: rowID
+```
+
+Each partition of the partitioned table is treated as a  independent physical table in the TiKV layer, with its own `partitionID` . Therefore, the index entries of the partitioned table are encoded as:
+
+```
+Key: tablePrefix{partitionID}_indexPrefixSep{indexID}_indexedColumnsValue
+Value: rowID
+```
+
+So, in the current table partition implementation, a partition and its indexes are encoded to the same key-value range. Or we can  say, TiDB only has local index.
+
+This design is compatible with the MySQL, but also makes TiDB have the same [constraint](https://dev.mysql.com/doc/refman/5.7/en/partitioning-limitations-partitioning-keys-unique-keys.html) as MySQL, which says:
+
+> Every unique key on the table must use every column in the table's partitioning expression.
+
+For example, here is a partitioned table:
+
+```sql
+CREATE TABLE t (
+    A int,
+    B int,
+    UNIQUE KEY (B)
+)
+PARTITION BY RANGE (A) (
+ PARTITION p0 VALUES LESS THAN (10),
+ PARTITION p1 VALUES LESS THAN (20),
+ PARTITION p2 VALUES LESS THAN (maxvalue)
+);
+```
+
+Then we insert 2 records into it.
+
+```sql
+INSERT INTO t (A, B) VALUES (1, 23);
+INSERT INTO t (A, B) VALUES (11, 23);
+```
+
+The first record will be inserted into `p0`, the second will be inserted into `p1`. The index on column `B` will create two entries which keys are:
+
+```
+p0_idxB_23
+p1_idxB_23
+```
+
+Because index entries is built on different partition, uniqueness of index is broken. So, In MySQL and current TiDB, these indexes are not allowed.
+
+## Proposal
+
+### Overview
+
+This document proposes to support global index for partitioned tables. Global index use `tableID` instead of `partitionID`  as the prefix of  index entries. Since all entries in the partitioned table use same prefix, it can ensure index entries is unique in whole table, so that removing the constraint in partitioned tables indexes.
+
+### Encoding
+
+#### Index key-value Layout
+
+We want global index entry keys begin with `TableID`, so we have add `PartitionID` to another place. New index encoding layout:
+
+```
+Key: tableID_indexID_ColumnValues
+Value: handle_partitionID
+```
+
+**Pros:** This key layout is compatible with current index key encoding.
+
+**Cons:**  Hard to drop and exchange partition, because index entries of a partition is not continuous.
+
+##### **Other Alternative:**
+
+```
+Key: tableID_partitionID_indexID_ColumnValues
+Value: handle
+```
+
+Or
+
+```
+Key: tableID_indexID_partitionID_ColumnValues
+```
+
+**Pros:** Easy to implement partition operators such as drop partition.
+
+**Cons:** Since `partitionID` is unknown when building query, it need scan all partitions of a table when handling point-get or index scan. Also, we need specific check when insert a row into table to keep index unique. It likes another implement of local index but not global index, So we reject it.
```

Comment:
```suggestion
**Cons:** Since `partitionID` is unknown when building a query, it needs to scan all partitions of a table when handling point-get or index scan. Also, we need special checks when inserting a row into a table to keep the index unique. It likes another implement of local index but not global index, So we reject it.
```